File: Leetcode Problems/287.py
from typing import List

class Solution:
    def findDuplicate(self, n: List[int]) -> int:
        slow = n[0]
        fast = n[0]

        while True:
            slow = n[slow]
            fast = n[n[fast]]
            if slow == fast:
                break

        slow = n[0]
        while slow != fast:
            slow = n[slow]
            fast = n[fast]
        
        return slow
    
if __name__ == '__main__':
    s = Solution()
    print(s.findDuplicate([1,3,4,2,2])) # 2

File: Leetcode Problems/287.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
    int findDuplicate(vector<int>& n) {
        int slow = n[0];
        int fast = n[0];

        do {
            slow = n[slow];
            fast = n[n[fast]];
        } while(slow != fast);

        slow = n[0];

        while(slow != fast) {
            slow = n[slow];
            fast = n[fast];
        }

        return slow;

    }
};

int main() {
    Solution s;
    vector<int> n = {1, 3, 4, 2, 2};
    cout << s.findDuplicate(n) << endl;
    return 0;
}